File: leetcode/338/Counting Bits.cpp
#include <iostream>
#include <vector>

using std::cin;
using std::cout;
using std::endl;
using std::vector;

class Solution
{
  public:
    vector<int> countBits(int num)
    {
        vector<int> result;
        for (int i = 0; i <= num; ++i)
        {
            int number = i;
            int count = 0;
            while (number)
            {
                if (number % 2)
                {
                    ++count;
                }
                number /= 2;
            }
            result.push_back(count);
        }
        return result;
    }

    void printOut(vector<int> result)
    {
        for (vector<int>::iterator beg = result.begin(); beg != result.end(); ++beg)
        {
            cout << *beg << ' ';
        }
        cout << endl;
    }
};

int main(int argc, char **argv)
{
    int num = 2;
    int size;
    Solution s = Solution();
    vector<int> result = s.countBits(num);
    s.printOut(result);
    num = 5;
    result = s.countBits(num);
    s.printOut(result);
    num = 10;
    result = s.countBits(num);
    s.printOut(result);
    return 0;
}

File: leetcode/338/Counting Bits.py
"""Counting Bits
Given a non negative integer number num. For every numbers i in the
range 0 ≤ i ≤ num calculate the number of 1's in their binary representation
and return them as an array.

Example 1:

Input: 2
Output: [0,1,1]

Example 2:

Input: 5
Output: [0,1,1,2,1,2]
Follow up:

    1. It is very easy to come up with a solution with run time
    O(n*sizeof(integer)). But can you do it in linear time O(n) /possibly in a
    single pass?

    2. Space complexity should be O(n).

    3. Can you do it like a boss? Do it without using any builtin function
    like __builtin_popcount in c++ or in any other language.
"""


class Solution1:
    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]

        Examples
        
            >>> s = Solution1()
            >>> num = 2
            >>> s.countBits(num)
            [0, 1, 1]
            >>> num = 5
            >>> s.countBits(num)
            [0, 1, 1, 2, 1, 2]
        """
        result = []
        for i in range(num + 1):
            result.append(bin(i).count('1'))
        return result


class Solution2:
    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]

        Examples
        
            >>> s = Solution2()
            >>> num = 2
            >>> s.countBits(num)
            [0, 1, 1]
            >>> num = 5
            >>> s.countBits(num)
            [0, 1, 1, 2, 1, 2]
        """
        result = []
        count = 0
        for i in range(num + 1):
            while i:
                if (i % 2):
                    count += 1
                i //= 2
            result.append(count)
            count = 0
        return result


if __name__ == "__main__":
    import doctest
    doctest.testmod(verbose=True)
